### PR TITLE
Refactor the validator

### PR DESCRIPTION
**Pull request by philectron**
I want to try out the GitHub app `WIP` that I added a few days ago. It won't allow merge until the `WIP` tag is taken out of the title. And yes, this branch is in fact a work-in-progress.

Note that this will be merged to the `merge` branch, not `master`.

### CODE REVIEW

**Comment by philectron** [comment]
I recommend clone this branch and test the new validator. You can test it directly in the Node console on your terminal. You only need to import `validation.js`, `models/plan.js`, and create a few objects in the Node console to test with. Don't forget to test the sanitizer, too

Note that this validator does not (or not yet) check the course compatibility of a submitted plan. I'd still like to merge this branch to the `merge` branch without integrating it because the task to integrate it to the existing API endpoints is a large one, which should be done on another branch. It should also be done by the person who wrote the current API endpoints (aka @silverware13), as the refactor will be a big one, and I don't feel too comfortable editing too much of someone else's code 😄. My schedule is quite tight as well...

**Comment by philectron** [comment]
@Mergifyio refresh

**Comment by philectron** [comment]
Alright! Looks like it can auto-merge from now on.

**Comment by philectron** [comment]
@silverware13 I'm ready.

**Comment by philectron** [comment]
My test sample:

File: utils/schemaValidation.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        "The user ID associated with this plan must be an integer at least " +
+        `${this.minValue}.`;
+    }
+  },
+  lastUpdated: {
+    required: true,
+    type: Type.timestamp,
+    getErrorMessage: function() {
+      return "Constraint violated: Invalid plan timestamp\n" +
+        "The plan timestamp must be in ISO 8601 format.";
```
Comment (silverware13) [inline utils/schemaValidation.js]: I may be misunderstanding the schema, but timestamp should not be required for the submission, the database will auto set the timestamp when a new submission is made (kind of like what you expect with planId).

File: utils/schemaValidation.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  status: {
+    required: true,
+    type: Type.integer,
+    minValue: 0,
+    maxValue: 4,
+    getErrorMessage: function() {
+      return "Constraint violated: Invalid plan status\n" +
+        "Plan status must be 0 (Rejected), 1 (Awaiting Student Changes) " +
+        "2 (Awaiting Reivew), 3 (Awaiting Final Review), or 4 (Accepted).";
+    }
```
Comment (silverware13) [inline utils/schemaValidation.js]: I may not be understanding this correctly, but a status should not be required when the plan is submitted. The plan will always have a default status, which then can be updated by advisors at a later time.

File: utils/schemaValidation.js (hunk 1 above)
Comment (philectron) [reply utils/schemaValidation.js]: You're right. I'm changing it to not required.

File: utils/schemaValidation.js (hunk 2 above)
Comment (philectron) [reply utils/schemaValidation.js]: I heard you this morning but forgot to update it :D. Thanks!

**Review by silverware13 — APPROVED** [review]

**Comment by philectron** [comment]
It did it! It auto-merged!

**Pull request merged**